pkg/status/combinedstatus-resolver.go

```diff
@@ -0,0 +1,470 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"fmt"
+	"sync"
+
+	runtime2 "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/kubestellar/kubestellar/api/control/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/abstract"
+	"github.com/kubestellar/kubestellar/pkg/binding"
+	controllisters "github.com/kubestellar/kubestellar/pkg/generated/listers/control/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+type CombinedStatusResolver interface {
+	// GenerateCombinedStatus generates a CombinedStatus object for the given
+	// binding name and workload object identifier.
+	// If no resolution is associated with the given combination, nil is returned.
+	GenerateCombinedStatus(bindingName string, objectIdentifier util.ObjectIdentifier) *v1alpha1.CombinedStatus
+
+	// CompareCombinedStatus compares the given CombinedStatus object with the
+	// one associated with the given binding name and workload object identifier.
+	// True is returned in case of a match, false otherwise.
+	CompareCombinedStatus(bindingName string, objectIdentifier util.ObjectIdentifier,
+		combinedStatus *v1alpha1.CombinedStatus) bool
+
+	// NoteBindingResolution notes a binding resolution for status collection.
+	//
+	// 1. If `deleted` is true, the associated combinedstatus resolutions are
+	// removed from memory. The same is done if a resolution no longer requires
+	// status collection.
+	//
+	// 2. Excessive combinedstatus resolutions are removed if they are no
+	// longer associated with the binding.
+	//
+	// 3. For every workload object associated with one or more
+	// statuscollectors, a combinedstatus resolution is created/updated.
+	// The update may involve adding or removing statuscollectors, and changing
+	// the set of destinations associated with the binding.
+	//
+	// The function uses the statuscollector lister and the cached workstatus
+	// objects to update internal state. The given array is assumed to contain
+	// all objects that may be relevant to the binding.
+	//
+	// The returned set contains the identifiers of combinedstatus objects
+	// that should be queued for syncing.
+	NoteBindingResolution(bindingName string, bindingResolution binding.Resolution, deleted bool,
+		workStatuses []*workStatus,
```

Comment:
There could be a lot of these that are irrelevant, right? Wouldn't it be more efficient to pass a function that returns the needed WorkStatus (which would be implemented by a fetch from the local cache of the WorkStatus informer)?

Reply:
Whoever calls NoteBindingResolution has enough information to limit the passed list to what is necessary. Unfortunately, at the moment, we can only filter fetching from the informer's cache by WDS name. WorkStatuses are labelled with the name of a BindingPolicy but I'm not sure how that addresses multiple BPs selecting the same objects.

Therefore, the fetching is at a higher level, and when consuming workstatuses there is filtering to ignore what's irrelevant.

Reply:
Remember that when I talk about an "informer" I mean a `SharedIndexInformer`. These thing support indexing. This controller could add an index that supports fetching the exact one WorkStatus that is needed at a given point in the code.

Reply:
Check this https://github.com/vMaroon/kubestellar/pull/11/files#r1642771147